blockly-rtc/src/WorkspaceClient.js

```diff
@@ -87,39 +78,46 @@ export default class WorkspaceClient {
-        };
+      this.beginWrite_();
+      try {
+        await writeEvents(this.inProgress[0]);
+        this.endWrite_(true);
+      } catch {
+        this.endWrite_(false);
+        throw Error('Failed to write to database.');
+      };
     };
 
     /**
      * Change status of WorkspaceClient in preparation for the network call.
-     * Set writeInProgress to true and move events from notSent to inProgress.
+     * Set writeInProgress to true, adds a LocalEntry to inProgress based on
+     * the events that were notSent, and clears the notSent array.
      * @private
      */
     beginWrite_() {
-        this.writeInProgress = true;
-        this.inProgress = this.inProgress.concat(this.notSent);
-        this.notSent = [];
+      this.writeInProgress = true;
+      const entryId = this.workspaceId.concat(this.counter);
+      this.counter +=1;
+      this.inProgress.push({
+        events: this.notSent,
+        entryId: entryId
+      }); 
+      this.notSent = [];
     };
-
+                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           
     /**
      * Change status of WorkspaceClient once network call completes.
-     * Change writeInProgress to true. If write was successful events remain in
-     * inProgress, otherwise the inProgress events are moved to the beginning of
-     * notSent.
+     * Change writeInProgress to true. If write was successful, the LocalEntry
+     * remains in inProgress, otherwise, the LocalEntry is deleted and the
+     * events move back to the front of notSent.
      * @param {boolean} success Indicates the success of the database write.
      * @private
      */
     endWrite_(success) {
         if (!success) {
-            this.notSent = this.inProgress.concat(this.notSent);
+            this.notSent = this.inProgress[0].events.concat(this.notSent);
```

Comment:
This could be dangerous in the unlikely event that the server received the request, the response was lost, and new events were added in the meantime. In that case, you would resend the update with the same id, but with extra events. The server would recognize it as the same update and throw it away, losing the new events.

Instead, you'll want to keep the in progress update as is and retry it on the next iteration.

Reply:
Thanks for pointing out this concern :)
I will create a task for implementing retry functionality and add it as a TODO.
I would like the focus of this PR to remain on the redesign of the storage schema, so I will address the TODO in a following PR.